operator/src/lib.rs

```diff
@@ -0,0 +1,608 @@
+                // create pod when 1) no hash found or 2) hash found but current pod count < spec pod count
+                if let Some(pods) = nodes.get(hash) {
+                    current_count = pods.len();
+                    if current_count > spec_pod_count {
+                        let pod = pods.get(0).unwrap();
+                        self.context.client.delete(pod).await?;
+                        info!("deleting {} pod '{}'", node_type.as_str(), Meta::name(pod));
+                    }
+                }
+
+                if current_count < spec_pod_count {
+                    let pod = self.create_pod(selector, node_type, hash).await?;
+                    self.create_config_maps(selector, node_type, hash).await?;
+                    info!("Creating {} pod '{}'", node_type.as_str(), Meta::name(&pod));
+                }
+            }
+        }
+
+        Ok(ReconcileFunctionAction::Continue)
+    }
+
+    // fn check_pods_ready(&self, hashed_pods: &HashMap<String, Vec<Pod>>) -> bool {
+    //     for pods in hashed_pods.values() {
+    //         for pod in pods {
+    //             // terminating
+    //             if self.has_deletion_stamp(pod) {
+    //                 return false;
+    //             }
+    //             if self.is_pod_running_and_ready(pod) {
+    //                 return false;
+    //             }
+    //         }
+    //     }
+    //     return true;
+    // }
+    //
+    // fn has_deletion_stamp(&self, pod: &Pod) -> bool {
+    //     if finalizer::has_deletion_stamp(pod) {
+    //         info!(
+    //             "SparkCluster {} is waiting for Pod [{}] to terminate",
+    //             self.context.log_name(),
+    //             Meta::name(pod)
+    //         );
+    //         return true;
+    //     }
+    //     return false;
+    // }
+    //
+    // fn is_pod_running_and_ready(&self, pod: &Pod) -> bool {
+    //     if !podutils::is_pod_running_and_ready(pod) {
+    //         info!(
+    //             "SparkCluster {} is waiting for Pod [{}] to be running and ready",
+    //             self.context.log_name(),
+    //             Meta::name(pod)
+    //         );
+    //         return false;
+    //     }
+    //     return true;
+    // }
+
+    async fn create_pod(
+        &self,
+        selector: &SparkNodeSelector,
+        node_type: &SparkNodeType,
+        hash: &str,
+    ) -> Result<Pod, Error> {
+        let pod = self.build_pod(selector, hash, node_type)?;
+        Ok(self.context.client.create(&pod).await?)
+    }
+
+    fn build_pod(
+        &self,
+        selector: &SparkNodeSelector,
+        hash: &str,
+        node_type: &SparkNodeType,
+    ) -> Result<Pod, Error> {
+        Ok(Pod {
+            metadata: self.build_pod_metadata(node_type, hash)?,
+            spec: Some(self.build_pod_spec(selector, node_type, hash)),
+            ..Pod::default()
+        })
+    }
+
+    fn build_pod_metadata(
```

Comment:
operator-rs has a method that you should be able to use: `metadata::build_metadata`